src/django_components/provide.py

```diff
@@ -102,24 +102,22 @@ def render(self, context: Context, name: str, **kwargs: Any) -> SafeString:
 
 
 def get_injected_context_var(
+    component_id: str,
     component_name: str,
-    context: Context,
     key: str,
     default: Optional[Any] = None,
 ) -> Any:
     """
     Retrieve a 'provided' field. The field MUST have been previously 'provided'
     by the component's ancestors using the `{% provide %}` template tag.
     """
-    # NOTE: For simplicity, we keep the provided values directly on the context.
-    # This plays nicely with Django's Context, which behaves like a stack, so "newer"
-    # values overshadow the "older" ones.
-    internal_key = _INJECT_CONTEXT_KEY_PREFIX + key
+    # NOTE: `component_provides` is defaultdict. Use `.get()` to avoid making an empty dictionary.
+    providers = component_provides.get(component_id)
```

Comment:
Previously the provided data was stored for the duration of the context.

But to make the provide/inject work even outside of rendering, the provided data is stored for the lifetime of the Component instance. Our Component instances may now outlive the Django's Context object.

As such, when we call `Component.inject()`, we now access the provided data using the component instance as the key.